crates/pessimistic-proof-utils/src/local_exit_tree.rs

```diff
@@ -0,0 +1,249 @@
+use std::fmt::Debug;
+
+use pessimistic_proof::local_exit_tree::hasher::Hasher;
+use pessimistic_proof::local_exit_tree::LocalExitTree;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+/// Represents a local exit tree as defined by the LxLy bridge.
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LocalExitTreeData<H, const TREE_DEPTH: usize = 32>
+where
+    H: Hasher,
+    H::Digest: Serialize + DeserializeOwned,
+{
+    /// The layers of the Merkle tree from bottom to top (i.e., the leaves are
+    /// in `layers[0]`)
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    layers: [Vec<H::Digest>; TREE_DEPTH],
+    /// `empty_hash_at_height[i]` is the root of an empty Merkle tree of depth
+    /// `i`.
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    empty_hash_at_height: [H::Digest; TREE_DEPTH],
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LETMerkleProof<H, const TREE_DEPTH: usize = 32>
+where
+    H: Hasher,
+    H::Digest: Serialize + DeserializeOwned,
+{
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    siblings: [H::Digest; TREE_DEPTH],
+}
+
+impl<H, const TREE_DEPTH: usize> LocalExitTreeData<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Default + Serialize + DeserializeOwned,
+{
+    /// Creates a new empty [`LocalExitTreeData`].
+    pub fn new() -> Self {
+        let mut empty_hash_at_height = [H::Digest::default(); TREE_DEPTH];
+        for height in 1..TREE_DEPTH {
+            empty_hash_at_height[height] = H::merge(
+                &empty_hash_at_height[height - 1],
+                &empty_hash_at_height[height - 1],
+            );
+        }
+        LocalExitTreeData {
+            layers: std::array::from_fn(|_| Vec::new()),
+            empty_hash_at_height,
+        }
+    }
+
+    /// Creates a new [`LocalExitTreeData`] and populates its leaves.
+    pub fn from_leaves(leaves: impl Iterator<Item = H::Digest>) -> Self {
```

Comment:
This could be turned into a `FromIterator` impl so the tree can be constructed with the `.collect()` iterator method.

(Also it is probably more efficient to bulk-construct the tree level by level but I'm not sure it's worth implementing. Depends on how often is such bulk construction used in production code which I'd guess maybe not very often.)

Reply:
Good idea for `FromIterator`, will implement it 👍 
Regarding the bulk-adding of leaves, I think it could be useful, but
1. It would be slightly annoying to implement optimally
2. This function should be extremely fast compared to other parts of the code so repeatedly calling `add_leaf`, although not optimal, should be more than good enough.

I can add a TODO for now and we can consider doing that in the future when optimizing.

Reply:
@wborgeaud For any TODO, can you create an associated ticket in the [V1 - Technical improvements](https://github.com/orgs/agglayer/projects/4?query=sort%3Aupdated-desc+is%3Aopen) part of the project to track down those improvements please?

Reply:
Sounds good, will do once this is merged 👍